### PR TITLE
Credentials 5: Machine account auth

### PR DESCRIPTION
**Pull request by ianwremmel**

### CODE REVIEW

File: packages/plugin-machine-account/src/machine-account.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
+ * @private
+ */
+
+import {SparkPlugin} from '@ciscospark/spark-core';
+
+/**
+ * Interacts with the deprecated /bots API on the conversation service. *THESE
+ * ARE NOT THE BOTS YOU'RE LOOKING FOR*. This package is only for helping with
```
Comment (Josh-Dykstra) [inline packages/plugin-machine-account/src/machine-account.js]: 💯

File: packages/plugin-machine-account/test/integration/spec/machine-account.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/**!
+ *
+ * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
+ * @
```
Comment (Josh-Dykstra) [inline packages/plugin-machine-account/test/integration/spec/machine-account.js]: private?

File: packages/plugin-machine-account/test/integration/spec/machine-account.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    describe(`#delete()`, () => {
+      describe(`when the current account owns the machine account`, () => {
+        let bot;
+        before(() => spark.machineAccount.create({
+          name: `spark-js-sdk-testbot-${uuid.v4()}`,
+          contactEmail: user.email
+        })
+          .then((b) => {
+            // not addign to bots array because we're going to delete it as part of the test
```
Comment (Josh-Dykstra) [inline packages/plugin-machine-account/test/integration/spec/machine-account.js]: typo

File: packages/plugin-machine-account/src/machine-account.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+   */
+  create(options) {
+    options = options || {};
+
+    if (!options.name) {
+      return Promise.reject(new Error(`\`options.name\` is required`));
+    }
+
+    if (!options.contactEmail) {
+      return Promise.reject(new Error(`\`options.contactEmail\` is required`));
```
Comment (Josh-Dykstra) [inline packages/plugin-machine-account/src/machine-account.js]: Unit test coverage? :)

File: packages/plugin-credentials/src/credentials.js [hunk 5]
```diff
@@ -218,6 +218,34 @@ const Credentials = SparkPlugin.extend({
...
 
+  /**
+   * Exchanges an orgId/password/name triple for an access token. If you're
+   * considering using this method, you're almost certainly interested in "bots"
+   * rather than "machine accounts". See the developer portal for more
+   * information.
+   */
+  @whileInFlight(`isAuthenticating`)
+  @oneFlight
+  @retry()
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/credentials.js]: image

File: packages/plugin-credentials/test/integration/spec/credentials.js [hunk 6]
```diff
@@ -64,12 +66,35 @@ describe(`plugin-credentials`, () => {
...
+        contactEmail: user.email
+      })
+        .then((b) => {
+          bot = b;
+          sparkBot = new CiscoSpark();
+        }));
+
+      afterEach(() => spark.machineAccount.delete(bot)
+        .catch(() => spark.machineAccount.delete(bot))
+        .catch(() => spark.machineAccount.delete(bot)));
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/test/integration/spec/credentials.js]: Third time's a charm?

File: packages/plugin-credentials/src/credentials.js [hunk 7]
```diff
@@ -246,6 +274,80 @@ const Credentials = SparkPlugin.extend({
...
+      return Promise.reject(new Error(`\`options.name\` is required`));
+    }
+
+    if (!options.password) {
+      return Promise.reject(new Error(`\`options.password\` is required`));
+    }
+
+    return this.spark.request({
+      method: `POST`,
+      uri: `{this.config.samlUrl}/${options.orgId}/v2/actions/GetBearerToken/invoke`,
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/credentials.js]: Seems like this is missing a `$` at the front. I don't think the string interpolation will turn that into a variable without it.

File: packages/plugin-machine-account/src/machine-account.js (hunk 4 above)
Comment (ianwremmel) [inline packages/plugin-machine-account/src/machine-account.js]: it's only here to enable tests for code that we'll probably never use

File: packages/plugin-credentials/src/credentials.js (hunk 7 above)
Comment (ianwremmel) [inline packages/plugin-credentials/src/credentials.js]: indeed

File: packages/plugin-credentials/test/integration/spec/credentials.js (hunk 6 above)
Comment (ianwremmel) [inline packages/plugin-credentials/test/integration/spec/credentials.js]: the service is a little flaky and retry wasn't behaving

**Pull request merged**